keps/sig-storage/20191125-bucket-provisioning.md

```diff
@@ -0,0 +1,757 @@
+1. `finalizers`: added by COSI to defer `BucketRequest` deletion until backend deletion succeeds.
+1. `protocol`: (required) specifies the desired protocol.  One of {“s3”, “gcs”, or “azureBlob”}.
+1. `bucketPrefix`: (optional) prefix prepended to a COSI-generated new bucket name, eg. “yosemite-photos-". If `bucketInstanceName` is supplied then `bucketPrefix` is ignored because the request is for access to an existing bucket.
+1. `bucketClassName`: (optional) name of the `BucketClass` used to provision this request. If omitted, a default bucket class matching the protocol is used. If the bucket class does not support the requested protocol, an error is logged and the request is retried. A `BucketClass` is required for both greenfield and brownfield requests since BCs support a list of allowed namespaces.
+1. `bucketInstanceName`: (optional) name of the cluster-wide `Bucket` instance. If blank, then COSI assumes this is a greenfield request an will fill in the name during the binding step. If defined by the user, then this names the `Bucket` instance created by the admin. There is no automation available to make this name known to the user.
+1. `bucketAvailable`: if true the bucket has been provisioned. If false then the bucket has not been provisioned and is unable to be accessed.
+
+#### Bucket
+
+A cluster-scoped custom resource representing the abstraction of a single backend bucket. A `Bucket` instance stores enough identifying information so that drivers can accurately target the backend object store (e.g. needed during a deletion process).  All of the associated bucket class fields are copied to the `Bucket`. Additionally, endpoint data returned by the driver is copied to the `Bucket` by the sidecar.
+
+For greenfield, COSI creates the `Bucket` based on values in the `BucketRequest` and `BucketClass`. For brownfield, an admin manually creates the `Bucket` and COSI copies bucket class fields, populates fields returned by the provisioner, and binds the `Bucket` to the `BucketAccess`.
+
+Since there is a 1:1 mapping between a BR and a `Bucket`, when multiple BRs request access to the same backend bucket, each separate `Bucket` points to the same backend bucket.
+
+```
+apiVersion: cosi.io/v1alpha1
+kind: Bucket
+metadata:
+  name: [1]
+  labels:
+    cosi.io/provisioner: [2]
+  finalizers:
+    - cosi.io/finalizer [3]
+spec:
+  provisioner: [4]
+  retentionPolicy: [5]
+  anonymousAccessMode: [6]
+  bucketClassName: [7]
+  bucketRequest: [8]
+    name:
+    namespace:
+  bucketAccessName: [9]
+  protocol: [10]
+    protocolSignature: ""
+    azureBlob:
+      containerName:
+      storageAccount:
+    s3:
+      version:
+      endpoint:
+      bucketName:
+      region:
+      signatureVersion:
+    gcs:
+      bucketName:
+      privateKeyName:
+      projectId:
+      serviceAccount:
+  allowedNamespaces: [11]
```

Comment:
Is allowedNamespaces still relevant now that every namespace has its own cluster-wide Bucket instance?

Reply:
Yes. _allowedNamespaces_ is defined in the BC and copied to the `Bucket` instance when COSI provisions a new backend bucket. This field lets an admin control which namespaces are allowed to request the creation of a bucket, or to request access to a bucket. A while back it was dropped from the KEP but community feedback was to add it back. As a note: a BC is immutable (like true for storage classes) but a `Bucket` instance can be mutated by the admin. So, _allowedNamespace_ in the `Bucket` could be changed to support additional namespaces or to remove a ns.

Reply:
I understand what it was meant to be before. But now that we changed B to be 1-1 with BR per user/app namespace, then what is the meaning of having allowedNamespace in every B that is off sync between all the B's that refer to the same backend bucket?

Reply:
@guymguym I understand your question better now: once we have a B it implicitly means that the user was allowed to create their BR in their ns and no other namespaces will using that B. Good point! I think we only need _allowedNamespaces_ in the BC and not replicated to the B.

Reply:
Continued in https://github.com/kubernetes/enhancements/pull/1383/files#r483796754